semantics/small-steps-test/test/Control/State/Transition/Examples/Modular/Base.hs

```diff
@@ -0,0 +1,359 @@
+  case admin of
+    OpenAccount withName withBalance -> do
+      Map.notMember withName accounting ?! AccountAlreadyExists withName
+      pure . flip inject st . Accounting @e $
+        Map.insert
+          withName
+          (Account @e withName withBalance)
+          accounting
+    CloseAccount withName -> do
+      accounting' <- case Map.lookup withName accounting of
+        Nothing ->
+          (failBecause $ NonExistingAccount withName)
+            >> pure accounting
+        Just x
+          | balance x /= mempty ->
+            (failBecause $ CloseNonEmptyAccount x) >> pure accounting
+        _ -> pure $ Map.delete withName accounting
+
+      pure $! inject (Accounting @e accounting') st
+    Withdraw fromAcnt amt ->
+      flip inject st . Accounting @e <$> case Map.lookup fromAcnt accounting of
+        Nothing ->
+          (failBecause $ NonExistingAccount fromAcnt)
+            >> pure accounting
+        Just x
+          | balance x < amt ->
+            (failBecause $ WithdrawInsufficientBalance x amt)
+              >> pure accounting
+        Just x ->
+          pure $
+            Map.insert
+              fromAcnt
+              (x {balance = balance x <> invert amt})
+              accounting
+
+data OPS
+
+data OPSPredicateFailure e
+  = TXFailure (TXPredicateFailure e)
+  | ADMINFailure (AdminPredicateFailure e)
+
+deriving stock instance Eq (OPSPredicateFailure Era_0)
+
+deriving stock instance Show (OPSPredicateFailure Era_0)
+
+instance STS (OPS) where
+  type Environment (OPS) = ()
+  type State (OPS) = (BT.Accounting Era_0)
+  type Signal (OPS) = [BT.Op Era_0]
+
+  type PredicateFailure (OPS) = OPSPredicateFailure Era_0
+
+  initialRules = []
+
+  transitionRules =
+    [processOps @(OPS) @TX @(ADMIN) @Era_0]
+
+instance Embed (TX) (OPS) where
+  wrapFailed = TXFailure
+
+instance Embed (ADMIN) (OPS) where
+  wrapFailed = ADMINFailure
+
+processOps ::
+  forall ops tx admin e.
+  ( HasLens (Signal ops) [Op e],
+    HasLens (State ops) (BT.Accounting e),
+    HasGetter (BT.Tx e) (Signal tx),
+    HasGetter (BT.Accounting e) (State tx),
+    HasGetter (Environment ops) (Environment tx),
+    HasGetter (BT.Admin e) (Signal admin),
+    HasGetter (BT.Accounting e) (State admin),
+    HasGetter (Environment ops) (Environment admin),
+    HasSetter (State ops) (State tx),
+    HasSetter (State ops) (State admin),
+    Embed tx ops,
+    Embed admin ops
+  ) =>
+  TransitionRule ops
+processOps = do
+  TRC (env, st, ops) <-
+    judgmentContext
+
+  case extract @_ @([Op e]) ops of
+    x : xs -> case x of
+      OpTx tx ->
+        do
+          txs <-
+            trans @tx
+              ( TRC
+                  ( extract env,
+                    extract $
+                      extract @_ @(BT.Accounting e)
+                        st,
+                    extract tx
+                  )
+              )
+          trans @ops (TRC (env, inject txs st, inject xs ops))
```

Comment:
Does the call to `trans` here means that `trans @ops` has to modify the entire `State tx`?

Reply:
Hmmm, no, since this is an `ops` transition rule `trans @ops` should modify `State ops`.